pkg/compose/compose_test: TestComposeCompare failed

pkg/compose/compose_test.TestComposeCompare [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_ComposeTestsBazel/20108434?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_ComposeTestsBazel/20108434?buildTab=artifacts#/) on master @ [7b9fd1a5c621d91b6bcfb6e04482a1a884e3bc24](https://github.com/cockroachdb/cockroach/commits/7b9fd1a5c621d91b6bcfb6e04482a1a884e3bc24):


```
        cockroach1-1  | *
        cockroach2-1  | CockroachDB node starting at 2025-07-06 05:15:12.130189217 +0000 UTC m=+1.257297789 (took 0.9s)
        cockroach2-1  | build:               CCL v25.3.0-alpha.4-dev-7b9fd1a5c621d91b6bcfb6e04482a1a884e3bc24 @ 2025/07/06 05:06:29 (go1.23.7 X:nocoverageredesign)
        cockroach2-1  | malloc_conf:         prof:false,prof_active:true,background_thread:true,thp:never,metadata_thp:disabled,narenas:32,dirty_decay_ms:2000,muzzy_decay_ms:0
        cockroach2-1  | webui:               http://cockroach2:8080
        cockroach2-1  | sql:                 postgresql://root@cockroach2:26257/defaultdb?sslmode=disable
        cockroach2-1  | sql (JDBC):          jdbc:postgresql://cockroach2:26257/defaultdb?sslmode=disable&user=root
        cockroach2-1  | RPC client flags:    /cockroach/cockroach <client cmd> --host=cockroach2:26257 --insecure
        cockroach2-1  | logs:                /cockroach/store/logs
        cockroach2-1  | temp dir:            /cockroach/store/cockroach-temp4225897659
        cockroach2-1  | external I/O path:   /cockroach/store/extern
        cockroach2-1  | store[0]:            path=/cockroach/store
        cockroach2-1  | clusterID:           a5a3e11a-a357-4d54-a9ce-8e0e721f9422
        cockroach2-1  | status:              initialized new cluster
        cockroach2-1  | nodeID:              1
        cockroach1-1  | CockroachDB node starting at 2025-07-06 05:15:12.188876218 +0000 UTC m=+1.315901851 (took 0.9s)
        cockroach1-1  | build:               CCL v25.3.0-alpha.4-dev-7b9fd1a5c621d91b6bcfb6e04482a1a884e3bc24 @ 2025/07/06 05:06:29 (go1.23.7 X:nocoverageredesign)
        cockroach1-1  | malloc_conf:         prof:false,prof_active:true,background_thread:true,thp:never,metadata_thp:disabled,narenas:32,dirty_decay_ms:2000,muzzy_decay_ms:0
        cockroach1-1  | webui:               http://cockroach1:8080
        cockroach1-1  | sql:                 postgresql://root@cockroach1:26257/defaultdb?sslmode=disable
        cockroach1-1  | sql (JDBC):          jdbc:postgresql://cockroach1:26257/defaultdb?sslmode=disable&user=root
        cockroach1-1  | RPC client flags:    /cockroach/cockroach <client cmd> --host=cockroach1:26257 --insecure
        cockroach1-1  | logs:                /cockroach/store/logs
        cockroach1-1  | temp dir:            /cockroach/store/cockroach-temp4024819686
        cockroach1-1  | external I/O path:   /cockroach/store/extern
        cockroach1-1  | store[0]:            path=/cockroach/store
        cockroach1-1  | clusterID:           eacb22af-091f-406f-b9ed-0dde13892647
        cockroach1-1  | status:              initialized new cluster
        cockroach1-1  | nodeID:              1
        test-1        |     compare_test.go:109: Checking connection to: cockroach2
        test-1        | === RUN   TestCompare/mutators
        test-1        |     compare_test.go:118: starting test: mutators
        test-1        |     compare_test.go:127: initializing connection: cockroach1
        test-1        |     compare_test.go:127: initializing connection: cockroach2
        test-1        |     compare_test.go:149: ALTER TABLE "table _1" ADD FOREIGN KEY (col1_0) REFERENCES table_2 ("c ol\n2'_7")
        test-1        |     compare_test.go:150: cockroach2: exec: ERROR: there is no unique constraint matching given keys for referenced table table_2 (SQLSTATE 23503)
        test-1        | --- FAIL: TestCompare (1.78s)
        test-1        |     --- FAIL: TestCompare/mutators (0.75s)
        test-1        | FAIL
        
�[Ktest-1 exited with code 1
        Aborting on container exit...
         Container compare-test-1  Stopping
         Container compare-test-1  Stopped
         Container compare-cockroach1-1  Stopping
         Container compare-cockroach2-1  Stopping
         Container compare-cockroach2-1  Stopped
         Container compare-cockroach1-1  Stopped
        
    compose_test.go:160: exit status 1
--- FAIL: TestComposeCompare (9.74s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestComposeCompare.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-52204

--- a/pkg/compose/compare/compare/compare_test.go
+++ b/pkg/compose/compare/compare/compare_test.go
@@ -93,7 +93,7 @@ func TestCompare(t *testing.T) {
 						randgen.ForeignKeyMutator,
 						randgen.ColumnFamilyMutator,
 						randgen.IndexStoringMutator,
-						randgen.PartialIndexMutator,
+						randgen.DupPartialIndexMutator,
 					},
 				},
 			},

--- a/pkg/sql/randgen/mutator.go
+++ b/pkg/sql/randgen/mutator.go
@@ -9,6 +9,7 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
+	"fmt"
 	"math/rand"
 	"regexp"
 	"sort"
@@ -43,6 +44,10 @@ var (
 	// PartialIndexMutator adds random partial index predicate expressions to
 	// indexes.
 	PartialIndexMutator MultiStatementMutation = partialIndexMutator
+
+	// DupPartialIndexMutator create a new random partial index based on an
+	// existing index that is selected.
+	DupPartialIndexMutator MultiStatementMutation = duplicatePartialIndexMutator
 
 	// PostgresMutator modifies strings such that they execute identically
 	// in both Postgres and Cockroach (however this mutator does not remove
@@ -1161,6 +1166,84 @@ func partialIndexMutator(rng *rand.Rand, stmts []tree.Statement) ([]tree.Stateme
 			}
 		}
 	}
+	return stmts, changed
+}
+
+// duplicatePartialIndexMutator is a mutations.MultiStatementMutator that will
+// copy a random index definition and make it a partial index. Unlike
+// partialIndexMutator, the existing definition will stay the same.
+func duplicatePartialIndexMutator(rng *rand.Rand, stmts []tree.Statement) ([]tree.Statement, bool) {
+	changed := false
+	tables := getTableInfoFromDDLStatements(stmts)
+	var newStmts []tree.Statement
+	for _, stmt := range stmts {
+		switch ast := stmt.(type) {
+		case *tree.CreateIndex:
+			info, ok := tables[ast.Table.ObjectName]
+			if !ok {
+				continue
+			}
+
+			// If the index is not already a partial index, make it a partial index
+			// with a 50% chance. Do not mutate an index that was created to satisfy a
+			// FK constraint.
+			if ast.Predicate == nil &&
+				!hasReferencingConstraint(info, ast.Columns) &&
+				rng.Intn(2) == 0 {
+				astString := tree.AsStringWithFlags(ast, tree.FmtParsableNumerics)
+				astCopy, err := parser.ParseOne(astString)
+				if err != nil {
+					panic(err)
+				}
+				createIndexCopy := astCopy.AST.(*tree.CreateIndex)
+				if len(createIndexCopy.Name) > 0 {
+					createIndexCopy.Name = tree.Name(fmt.Sprintf("%s_partial", createIndexCopy.Name))
+				}
+				tn := tree.MakeUnqualifiedTableName(createIndexCopy.Table.ObjectName)
+				createIndexCopy.Predicate = randPartialIndexPredicateFromCols(rng, info.columnsTableDefs, &tn)
+				changed = true
+				newStmts = append(newStmts, createIndexCopy)
+			}
+		case *tree.CreateTable:
+			info, ok := tables[ast.Table.ObjectName]
+			if !ok {
+				panic("table info could not be found")
+			}
+			var newDefs []tree.TableDef
+			for _, def := range ast.Defs {
+				var idx *tree.IndexTableDef
+				switch defType := def.(type) {
+				case *tree.IndexTableDef:
+					idx = defType
+				case *tree.UniqueConstraintTableDef:
+					if !defType.PrimaryKey && !defType.WithoutIndex {
+						idx = &defType.IndexTableDef
+					}
+				}
+
+				if idx == nil {
+					continue
+				}
+
+				// If the index is not already a partial index, make it a partial
+				// index with a 50% chance.
+				if idx.Predicate == nil &&
+					!hasReferencingConstraint(info, idx.Columns) &&
+					rng.Intn(2) == 0 {
+					idxCopy := *idx
+					if len(idxCopy.Name) > 0 {
+						idxCopy.Name = tree.Name(fmt.Sprintf("%s_partial", idxCopy.Name))
+					}
+					tn := tree.MakeUnqualifiedTableName(ast.Table.ObjectName)
+					idxCopy.Predicate = randPartialIndexPredicateFromCols(rng, info.columnsTableDefs, &tn)
+					changed = true
+					newDefs = append(newDefs, &idxCopy)
+				}
+			}
+			ast.Defs = append(ast.Defs, newDefs...)
+		}
+	}
+	stmts = append(stmts, newStmts...)
 	return stmts, changed
 }
 

